src/basicEventScheduler.py

```diff
@@ -0,0 +1,108 @@
+  """
+  def __init__(self):
+    """
+      Constructor
+      @ In, None
+      @ Out, None
+    """
+    ExternalModelPluginBase.__init__(self)
+
+  def initialize(self, container, runInfoDict, inputFiles):
+    """
+      Method to initialize the Basic Event Scheduler model
+      @ In, container, object, self-like object where all the variables can be stored
+      @ In, runInfoDict, dict, dictionary containing all the RunInfo parameters (XML node <RunInfo>)
+      @ In, inputFiles, list, list of input files (if any)
+      @ Out, None
+    """
+
+  def _readMoreXML(self, container, xmlNode):
+    """
+      Method to read the portion of the XML that belongs to the Basic Event Scheduler model
+      @ In, container, object, self-like object where all the variables can be stored
+      @ In, xmlNode, xml.etree.ElementTree.Element, XML node that needs to be read
+      @ Out, None
+    """
+    container.basicEvents = {}
+
+    for child in xmlNode:
+      if child.tag == 'BE':
+        container.basicEvents[child.text.strip()] = [child.get('tin'),child.get('tfin')]
+      elif child.tag == 'timeID':
+        container.timeID = child.text.strip()
+      elif child.tag == 'variables':
+        variables = [str(var.strip()) for var in child.text.split(",")]
+      else:
+        raise IOError("basicEventScheduler: xml node " + str(child.tag) + " is not allowed")
+
+  def run(self, container, inputs):
+    """
+      This method generate an historySet from the a pointSet which contains initial and final time of the
+      basic events
+      @ In, inputDataset, dict, dictionary of inputs from RAVEN
+      @ In, container, object, self-like object where all the variables can be stored
+      @ Out, basicEventHistorySet, Dataset, xarray dataset which contains time series for each basic event
+    """
+    dataDict = {}
+    for key in container.basicEvents.keys():
+      dataDict[key] = []
+      dataDict[key].append(inputs[container.basicEvents[key][0]][0])
+      dataDict[key].append(inputs[container.basicEvents[key][1]][0])
+
+    inputDataset = pd.DataFrame.from_dict(dataDict, orient='index',columns=['tin', 'tfin'])
+
+    timeArray = np.concatenate([inputDataset['tin'],inputDataset['tfin']])
+    timeArraySorted = np.sort(timeArray,axis=0)
+    timeArrayCleaned = np.unique(timeArraySorted)
+
+    keys = list(container.basicEvents.keys())
+    dataVars={}
+    for key in keys:
+      dataVars[key]=(['RAVEN_sample_ID',container.timeID],np.zeros((1,timeArrayCleaned.shape[0])))
+
+    basicEventHistorySet = xr.Dataset(data_vars = dataVars,
+                                      coords    = dict(time=timeArrayCleaned,
+                                      RAVEN_sample_ID=np.zeros(1)))
+
+    for key in container.basicEvents.keys():
+      tin  = inputs[container.basicEvents[key][0]][0]
+      tend = inputs[container.basicEvents[key][1]][0]
+      indexes = np.where(np.logical_and(timeArrayCleaned>tin,timeArrayCleaned<=tend))
```

Comment:
why 'timeArrayCleaned>tin', but not 'timeArrayCleaned>=tin'?

Reply:
This is due to to the conversion from interval-based data to instant-based data (i.e., the history set). The convention here is that a logical value set to 1 at a specific time instant implies that the basic event was actually True from the previous time instant

Reply:
I see your point, it is better to clarify it in the user manual.

Reply:
Agree, added text.

Reply:
Based on the definition, 'tin' is the initial time for the basic event. Based on the calculation ```np.where(np.logical_and(timeArrayCleaned>tin,timeArrayCleaned<=tend))```, when the time instant is `tin`, the value are set to 0. Does this mean the event are still the same as before at `tin`, and does not change state at `tin`? Only the time large than `tin`, the basic event are actually changing state. Do I state correctly? @mandd